Source/DafnyDriver/CliCompilation.cs

```diff
@@ -192,70 +194,163 @@ public class CliCompilation {
 
       var canVerifies = resolution.CanVerifies?.DistinctBy(v => v.Tok).ToList();
 
+      var verifiedAssertions = false;
       if (canVerifies != null) {
-        canVerifies = FilterCanVerifies(canVerifies);
+        canVerifies = FilterCanVerifies(canVerifies, out var line);
+        verifiedAssertions = line != null;
 
         var orderedCanVerifies = canVerifies.OrderBy(v => v.Tok.pos).ToList();
         foreach (var canVerify in orderedCanVerifies) {
-          canVerifyResults[canVerify] = new CliCanVerifyResults();
-          await Compilation.VerifyCanVerify(canVerify, false);
+          var results = new CliCanVerifyResults();
+          canVerifyResults[canVerify] = results;
+          if (line != null) {
+            results.TaskFilter = t => KeepVerificationTask(t, line.Value);
+          }
+
+          var queueVerificationTask = Compilation.VerifyCanVerify(canVerify, results.TaskFilter);
+          var tasks = new List<Task>() { queueVerificationTask };
+          var timeLimitSeconds = TimeSpan.FromSeconds(options.Get(BoogieOptionBag.VerificationTimeLimit));
+          if (timeLimitSeconds.Seconds != 0) {
+            tasks.Add(Task.Delay(timeLimitSeconds));
+          }
+          await Task.WhenAny(tasks);
+          if (!queueVerificationTask.IsCompleted) {
+            Compilation.Reporter.Error(MessageSource.Verifier, canVerify.Tok,
+              "Dafny encountered an internal error during verification. Please report it at <https://github.com/dafny-lang/dafny/issues>.\n");
+            break;
+          }
+          await queueVerificationTask;
         }
 
         foreach (var canVerify in orderedCanVerifies) {
           var results = canVerifyResults[canVerify];
           try {
+            var timeLimitSeconds = TimeSpan.FromSeconds(options.Get(BoogieOptionBag.VerificationTimeLimit));
+            var tasks = new List<Task>() { results.Finished.Task };
+            if (timeLimitSeconds.Seconds != 0) {
+              tasks.Add(Task.Delay(timeLimitSeconds));
+            }
+            await Task.WhenAny(tasks);
+            if (!results.Finished.Task.IsCompleted) {
+              Compilation.Reporter.Error(MessageSource.Verifier, canVerify.Tok,
+                "Dafny encountered an internal error during verification. Please report it at <https://github.com/dafny-lang/dafny/issues>.\n");
```

Comment:
Same string above, let's piggy-back on this PR, could you please refactor it to a constant?